### ISSUE TITLE
Add default output file name (input_name.jxl) without an OUTPUT argument

### ISSUE THREAD

**Issue by eugenesvk** [libjxl/libjxl#1796]
**Is your feature request related to a problem? Please describe.**
Currently quick cli file conversions are a bit cumbersome as they require you to pass an explicit output file name argument for no good reason ;)

**Describe the solution you'd like**
I'd prefer the default output file name to be chosen by the cjxl tool automatically by just changing the file extension of the output file
`cjxl input.jpg` should work as is (without an `OUTPUT` argument) and produce a lossless `input.jxl` — very convenient!
(and then the benchmarking folks would need to pass another flag instead for the cjxl tool not to create a file)

**Describe alternatives you've considered**
Creating some shell shim that would parse the `INPUT` path, get the file name, replace the extension, and pass the output file name

**Additional context**
None

**Comment by mo271** [comment]
One problem with the suggested behavior would be to accidentally overwrite a file `input.jxl`, which is something that we don't want. However I agree that having a `--benchmarking` or `--no-output` flag nonetheless might make sense; encoding without giving a file name would then fail if no such flag is provided. This would make sure that the user not assumes that an `input.jxl` was written when we run `cjxl input.jpg` is run.

**Comment by mo271** [comment]
There is also some other behavior found in some tools, like writing output to stdout in case no file name is given.
I'd propose keeping it simple. For now the code in the pending pull request emits a warning when there is no output file given, to make clear that the user is expected to provide one and the "benchmarking folks" need top pass another flag to allow having no output file as you suggested.

**Comment by eugenesvk** [comment]
> There is also some other behavior found in some tools, like writing output to stdout in case no file name is given

Yeah, that's the worst default

> I'd propose keeping it simple

That's what my proposal is about — simplicity (to the user) — by **not** forcing users of the most **common** workflow (a quick cli conversion) to pass any file output names and (due to your good defaults of visually lossless `-d 1`) not even pass any encoding options!
Warnings etc. complicate this

**Comment by jendalinda** [comment]
Writing a small script that will create the output file name for you is not that hard. You may also add other options like effort or so.

**Comment by eugenesvk** [comment]
> not that hard

yep, but it's harder than it needs to be, so why? Check this user-friendly approach [cavif-rs](https://github.com/kornelski/cavif-rs) that has simple quality/speed arguments and also auto-saves (but not auto-overwrites without a flag) vs the much more involved [avifenc](https://github.com/AOMediaCodec/libavif) alternative

> You may also add other options like effort or so

but the defaults are good (props to the libjxl devs!), so most of the time I'd not need to go through the effort of tweaking the effort ;)

**Comment by mo271** [comment]
leaving open to re-consider adding the automatic writing

**Comment by mo271** [comment]
Here's another problem I see with automatic writing. Consider a user happily converting a file with 
`cjxl image.jpg` 
receiving `image.jxl`. Then they run `cjxl another_image.jpg` and get `another_image.jxl`.
 And then they want to do it again with different encoder settings, and instead of writing the command twice they might be tempted to do 
`cjxl -d 0.9 *.jpg`, 
They get the info that `--overwrite` is needed, and pass it. But this could lead to globbing expanding two files, and then a command like `cjxl -d 0.9 --overwrite image.jpg image2.jpg'` is run, thereby overwriting their original file. 

While I see potential problems here, what are the cases where it is "harder than it needs to be", i.e. when you write some script to batch convert files, won't you almost certainly already have access to the filename in question already? For instance a scenario I could image is this:

```bash
for file in *.jpg; do cjxl $file ${file%.jpg}.jxl; done
```
and I don't think it is much more involved than:
```bash
for file in *.jpg; do cjxl $file; done
```
Also there is more ambiguity, e.g. if a user has in image named `image.JPG`, would they perhaps expect `image.JXL`? I guess that argument could at least been made.

However, I don't think the choice made in cavif-rs is terrible, but I do think that this is the exception rather the the norm looking at various image encoders.

**Comment by eugenesvk** [comment]
> Here's another problem I see with automatic writing. Consider a user happily converting a file with `cjxl image.jpg` receiving `image.jxl`. Then they run `cjxl another_image.jpg` and get `another_image.jxl`. And then they want to do it again with different encoder settings, and instead of writing the command twice they might be tempted to do `cjxl -d 0.9 *.jpg`, They get the info that `--overwrite` is needed, and pass it. But this could lead to globbing expanding two files, and then a command like `cjxl -d 0.9 --overwrite image.jpg image2.jpg'` is run, thereby overwriting their original file.

But this `cjxl -d 0.9 *.jpg` globbing command will be translated to
`cjxl -d 0.9 --overwrite image.jpg `  → `image.jxl`
`cjxl -d 0.9 --overwrite image2.jpg` → `image2.jxl`
So no original is overwritten? (checked with cavif, it simply creates two separate `.avif` files)

Also, I'd actually force the extensions — the encoder should always add the `.jxl` extension instead of forcing the user into more potential mistakes like `cjxl -d 0.9 --overwrite image.jpg image2.jpg'`. What is the proper workflow that would require you to force a wrong extension for the output?


> While I see potential problems here, what are the cases where it is "harder than it needs to be", i.e. when you write some script to batch convert files, won't you almost certainly already have access to the filename in question already? For instance a scenario I could image is this:
> 
> ```shell
> for file in *.jpg; do cjxl $file ${file%.jpg}.jxl; done
> ```
> 
> and I don't think it is much more involved than:
> 
> ```shell
> for file in *.jpg; do cjxl $file; done
> ```

But it is more involved! Before switching to a more sane shell, I'd have had to google that cryptic bash incantation that removes the file extension. Also, why would you write `for file...` instead of the more convenient `cjxl *.jpg` from your first example — this one surely is less involved than having to parse file names/extensions!


> Also there is more ambiguity, e.g. if a user has in image named `image.JPG`, would they perhaps expect `image.JXL`? I guess that argument could at least been made.

Nope, it's highly unlikely that the user would want to preserve this `file_normal.EXTUPPER` mistake from some other poorly coded app that produced/touched this image in the past. Or what is the actual argument for the opposite?

> this is the exception rather the the norm looking at various image encoders

yes, good defaults are the exception everywhere; and image encoders are no exception!

**Comment by mo271** [comment]
> But this `cjxl -d 0.9 *.jpg` globbing command will be translated to
> `cjxl -d 0.9 --overwrite image.jpg ` → `image.jxl`
> `cjxl -d 0.9 --overwrite image2.jpg` → `image2.jxl`
> So no original is overwritten? (checked with cavif, it simply creates two separate `.avif` files)

I checked the cavif behavior, and it seems that it is translated to 
`cavif image.jpg image2.jpg`, which is fine for cavif, because it expects and `-o` in front of the output file name and does the think you expect when no `-o` flag is provided.

To understand your feature request better, do you advocate for having an additional `-o` for the output file? I sometimes find this a little cumbersome having to look up in tools if it is `-o` or `-outfile`, like it is in djpeg or pdfjam for example.

**Comment by eugenesvk** [comment]
> I checked the cavif behavior, and it seems that it is translated to `cavif image.jpg image2.jpg`, which is fine for cavif, because it expects and `-o` in front of the output file name and does the think you expect when no `-o` flag is provided.

You're correct — you'd need curly brace expansion to expand into multiple commands, so without an `-o` and forcing an extension it's indeed an issue, thanks for pointing that out

> To understand your feature request better, do you advocate for having an additional -o for the output file? 

Yes, I'd advocate for an explicit `-o` flag to set the output file name explicitly as I'm certain that the usability of simple `cjxl *.jpg` globbing commands is way more important than having to add an extra flag if you're anyway specifying the output name.

Also, I'd like to reiterate the proposal to force the output files to have `.jxl` extension — even with the `-o` flag it's the right thing to do (and more convenient), though maybe that'd require an extra check to avoid double `.jxl.jxl` in case the user passed `.jxl`

> I sometimes find this a little cumbersome having to look up in tools if it is -o or -outfile, like it is in djpeg or pdfjam for example

Not more cumbersome than having to look up whether it's `-o` or no flag :) (and also less dangerous) (though `-outfile` is again one of those bad defaults, so here the fault lies with djpeg)
